### ISSUE TITLE
Should pruning leave singletons?

### ISSUE THREAD

**Issue by Anaphory** [glottobank/python-newick#13]
I am using newick through beastling.
Because my tree and my data don't match, the tree is pruned. Currently, pruning `A` from
`((A:1,B:1):1,C:2)` results in `((B:1):1,C:2)`, and `resolve_polytomies` does not change the result.

Should converting this tree to `(B:2,C2)` be the responsibility of `prune` or `resolve_polytomies` method, optionally of either of those, or of a new `remove_singletons` method?

**Comment by xrotwang** [comment]
I think this should be the responsibility of `resolve_polytomies`. Otherwise pruning would always have to include a `resolve_polytomies` step - which doesn't make sense, if the tree was not resolved before pruning to begin with.

**Comment by Anaphory** [comment]
I'm not good at reading code today:
There actually is a `remove_redundant_nodes` already, and I think I may have stared at it before wondering about singleton roots, I'll just add a call to that method at the end of `resolve_polytomies` – should it be forced, optional with default being called or optional with default being not called?

**Comment by Anaphory** [comment]
WAAAIT. Beastling _does_ call `remove_redundant_nodes` and _then_ `remove_polytomies`, and _still_ produces non-binary trees. I need to investigate further and give an example if it does turn out to be `newick`'s fault.

**Comment by Anaphory** [comment]
Ah, it looks like the problem is the following:
Pruning `A` and `B` from `((A:1,B:1):1,C:2)` results in `(:1,C:2)`, as is somewhat consistent, but not very helpful. Calling `remove_redundant_nodes` or `resolve_polytomies` _afterwards_ works as intended, because the resulting tree is binary.

How should this be fixed? It seems to me that the cleanest way would be for `prune` to _not add redundant nodes_ unless specifically asked to.

**Comment by LinguList** [comment]
Hi @Anaphory, I'd really like to join in the discussion, since the newick repo plays also an important role for future [LingPy](http://github.com/lingpy/lingpy) releases, which will, hopefully also offer exchange with beastling, yet it would help myself to understand this much more clearly, if you gave a little toy example, so I could replicate it (just in the issue, some command-line examples), and see to which degree I can help the discussion.

**Comment by Anaphory** [comment]
Curious. Building a minimal example is a worthwhile task here, because I thought the following would be one, but isn't:

Using [restrict_languages.txt](https://github.com/glottobank/python-newick/files/245391/restrict_languages.txt) with the command

```
$ cd $YOUR_BEASTLING_ROOT/test/..
$ beastling  tests/configs/basic.conf tests/configs/starting_tree.conf restrict_languages.txt -o -
```

I expected to see `(aiw,((aas:0.0,kbt:0.0):0.0))` in the output, but I actually get `(aiw,(aas:0.0,kbt:0.0))` as intended. Time to clone my own configuration and dig deeper…

**Comment by LinguList** [comment]
Okay, I guess, we need to find some newick examples, so we can trace, where we have problems with polytomies and the like. Having more cases for testing would generally be useful. I should make my homework and provide some larger tree examples both with polytomies and without.

**Comment by Anaphory** [comment]
This just wrong, and is the source of my problem.

```
>>> R = newick.loads("((A:1,B:1):1,C:1)")[0]
>>> R.prune_by_names(["A"])
>>> R.newick
'((B:1):1,C:1)'
>>> R.remove_redundant_nodes(True)
>>> R.newick
'(:1,C:1)'
```

**Comment by LinguList** [comment]
I think we can close this issue now, since the other issue #14 and my proposed PR-fix #15 would also solve this one. If pruning leaves singletons, one calls remove_redundant_nodes, and the problem will be solved.
